kafka.go

```diff
@@ -0,0 +1,131 @@
+package eventlogger
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/IBM/sarama"
+)
+
+type KafkaSink struct {
+	// Format specifies the format the []byte representation is formatted in
+	// Defaults to JSONFormat
+	Format string
+
+	// Topic specifies the topic the event should be written to.
+	Topic string
+
+	// Brokers is a list of Kafka brokers addresses.
+	Brokers []string
+
+	// The maximum permitted size of a message (defaults to 1000000). Should be
+	// set equal to or smaller than the broker's `message.max.bytes`.
+	MaxMessageBytes int
+
+	// The following config options control how often messages are batched up and
+	// sent to the broker. By default, messages are sent as fast as possible, and
+	// all messages received while the current batch is in-flight are placed
+	// into the subsequent batch.
+	Flush struct {
+		// The best-effort number of bytes needed to trigger a flush. Use the
+		// global sarama.MaxRequestSize to set a hard upper limit.
+		Bytes int
+		// The best-effort number of messages needed to trigger a flush. Use
+		// `MaxMessages` to set a hard upper limit.
+		Messages int
+		// The best-effort frequency of flushes. Equivalent to
+		// `queue.buffering.max.ms` setting of JVM producer.
+		Frequency time.Duration
+		// The maximum number of messages the producer will send in a single
+		// broker request. Defaults to 0 for unlimited. Similar to
+		// `queue.buffering.max.messages` in the JVM producer.
+		MaxMessages int
+	}
+
+	Retry struct {
+		// The total number of times to retry sending a message (default 3).
+		// Similar to the `message.send.max.retries` setting of the JVM producer.
+		Max int
+		// How long to wait for the cluster to settle between retries
+		// (default 100ms). Similar to the `retry.backoff.ms` setting of the
+		// JVM producer.
+		Backoff time.Duration
+	}
+}
+
+// Type describes the type of the node as a Sink.
+func (_ *KafkaSink) Type() NodeType {
+	return NodeTypeSink
+}
+
+func (ks *KafkaSink) Process(_ context.Context, e *Event) (*Event, error) {
+	format := ks.Format
+	if format == "" {
+		format = JSONFormat
+	}
+
+	val, ok := e.Format(format)
+	if !ok {
+		return nil, errors.New("event was not marshaled")
+	}
+
+	c := ks.parseConfig()
+	producer, err := sarama.NewSyncProducer(ks.Brokers, c)
```

Comment:
Same here, I think we want to be careful here as with a high rate of logs we'll be creating and closing lots of connections